api/src/HttpCache/PurgeHttpCacheListener.php

```diff
@@ -0,0 +1,255 @@
+        $uow = $em->getUnitOfWork();
+
+        foreach ($uow->getScheduledEntityInsertions() as $entity) {
+            $this->gatherResourceTags($entity);
+            $this->gatherRelationTags($em, $entity);
+        }
+
+        foreach ($uow->getScheduledEntityUpdates() as $entity) {
+            $originalEntity = $this->getOriginalEntity($entity, $em);
+            $this->addTagForItem($entity);
+            $this->gatherResourceTags($entity, $originalEntity);
+        }
+
+        foreach ($uow->getScheduledEntityDeletions() as $entity) {
+            $originalEntity = $this->getOriginalEntity($entity, $em);
+            $this->addTagForItem($originalEntity);
+            $this->gatherResourceTags($originalEntity);
+            $this->gatherRelationTags($em, $originalEntity);
+        }
+    }
+
+    /**
+     * Purges tags collected during this request, and clears the tag list.
+     */
+    public function postFlush(): void {
+        $this->cacheManager->flush();
+    }
+
+    /**
+     * Computes the original state of the entity based on the current entity and on the changeset.
+     */
+    private function getOriginalEntity($entity, $em) {
+        $uow = $em->getUnitOfWork();
+        $changeSet = $uow->getEntityChangeSet($entity);
+        $classMetadata = $em->getClassMetadata(ClassUtils::getClass($entity));
+
+        $originalEntity = clone $entity;
+        $em->detach($originalEntity);
+        foreach ($changeSet as $key => $value) {
+            $classMetadata->setFieldValue($originalEntity, $key, $value[0]);
+        }
+
+        return $originalEntity;
+    }
+
+    /**
+     * Purges all collections (GetCollection operations), in which entity is listed on top level.
+     *
+     * If oldEntity is provided, purge is only done if the IRI of the collection has changed
+     * (e.g. for updating period on a ScheduleEntry and the IRI changes from /periods/1/schedule_entries to /periods/2/schedule_entries)
+     */
+    private function gatherResourceTags(object $entity, ?object $oldEntity = null): void {
+        try {
+            $resourceClass = $this->resourceClassResolver->getResourceClass($entity);
+            $resourceMetadataCollection = $this->resourceMetadataCollectionFactory->create($resourceClass);
+            $resourceIterator = $resourceMetadataCollection->getIterator();
+            while ($resourceIterator->valid()) {
+                /** @var ApiResource $metadata */
+                $metadata = $resourceIterator->current();
+
+                foreach ($metadata->getOperations() ?? [] as $operation) {
+                    if ($operation instanceof GetCollection) {
+                        $this->invalidateCollection($operation, $entity, $oldEntity);
+                    }
+                }
+                $resourceIterator->next();
+            }
+        } catch (InvalidArgumentException|OperationNotFoundException) {
+        }
+    }
```

Comment:
Do you know in which case the exceptions occur? 
Is it intentional that the `while`-loop is aborted in the case of an exception? 
Or should the exceptions be caught within the `while`-loop?

Reply:
Hmm, good question. Originally this comes from the upstream implementation of `PurgeHttpCacheListener`.

I think there, the `OperationNotFoundException` could occur, if a resource has no `GetCollection` operation implemented. But as we're now looping over the operations instead of trying to retrieve a specific operation, this exception should never occur.

I think the better option would be to remove the try/catch block here to not suppress any purging by mistake. If we have an edge case that is not covered properly, then at least it pops up in the tests or in Sentry eventually.

Would you agree?

Reply:
Commit reference: https://github.com/ecamp/ecamp3/pull/3610/commits/dd199d70b59eefc95ca7c73a705444972e45648e

Reply:
> Would you agree?

Yes